### ISSUE TITLE
deflate implementation in standard library

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#213]
These compression and decompression functions are so common that we should have them ship with the compiler. We'll probably need to depend on them for the package manager.

**Comment by itsmontoya** [comment]
Commenting here since Github won't let me be assigned.

I'll try to look into this asap

**Comment by daurnimator** [comment]
@itsmontoya @schmee did either of you get anywhere on this issue?

**Comment by devzero** [comment]
I didn't write it, but I found this: [https://github.com/tiehuis/zig-deflate](https://github.com/tiehuis/zig-deflate) that looks pretty far along

**Comment by emekoi** [comment]
the implementation that is based on isn't designed for speed.

**Comment by schmee** [comment]
@daurnimator I'm messing around with it but no promises about when/if I'll be done. So if someone else wants to work on this please feel free! 🙂

**Comment by daurnimator** [comment]
Once this is done it should be used so that we can store the glibc data compressed (see https://github.com/ziglang/zig/pull/2847#issuecomment-509276642 )

**Comment by andrewrk** [comment]
This is a really nice article on the subject: https://www.hanshq.net/zip.html

**Comment by mlarouche** [comment]
I'll writing one for PNG support in zigimg

**Comment by iamgreaser** [comment]
Currently working on an implementation myself. I've got decompression working (it's about 1/4 of the speed of gunzip although it doesn't write to an actual output file yet), and I started working on a compressor yesterday, although that hasn't been pushed yet.

https://github.com/iamgreaser/deflate-for-zig/

I can be found on #zig under the name GreaseMonkey, if you have any suggestions then let me know.

**Comment by mlarouche** [comment]
My current progress is here: https://github.com/mlarouche/zigimg/blob/ReadPNGSupport/src/compression/deflate.zig Haven't mesured performance yet

**Comment by iamgreaser** [comment]
OK, so we've had 3 decompressors so far, of which the earlier one appears to be a port of a C implementation, and the other two happen to both work on Zig 0.6.0.

As far as performance is concerned, it'll really only be slow if you aren't buffering your input. If you're using the `InStream` interface, then you can plug in a `BufferedInStream` and that'll magically fix the worst of it.

If this should be in the standard library, then what API should we provide? And what primitives would we want in the standard library to make this work?

I'm mostly thinking from the perspective of 

Here's what I'm thinking in terms of an API:

```zig
pub fn DeflateInStream(comptime TypeInStream: type) type { ... }
pub fn GzipInStream(comptime TypeInStream: type) type { ... } // FIXME: What capitalisation should we use?
pub fn ZlibInStream(comptime TypeInStream: type) type { ... }
```

And as for primitives, we do have `BufferedInStream`, `BitInStream`, and `hash.Crc32` already (daurnimator pointed some of these out on IRC), but we're likely to need something like these to become a thing (names and interfaces TBD):

```zig
pub fn CanonicalHuffmanReader(... TODO ..., comptime out_type: type, max_len: usize) type { ... }
pub fn LzSlidingWindow(... TODO ..., comptime size: usize) type { ... }
```

Thoughts?

**Comment by mlarouche** [comment]
Looks good to me :)

**Comment by Rocknest** [comment]
@andrewrk please reopen since INFLATE is not yet implemented

**Comment by andrewrk** [comment]
`std.compress.deflate.inflateStream`

**Comment by Rocknest** [comment]
I mean DEFLATE

**Comment by andrewrk** [comment]
ahh I see, my mistake.

**Comment by gsquire** [comment]
Would you be interested in including a [snappy](https://github.com/google/snappy) encoder and decoder? I implemented the block format [here](https://github.com/gsquire/zig-snappy). It needs some more tests and polish but it's in a working state as of now.

**Comment by Mouvedia** [comment]
Maybe we could have a GitHub label for https://github.com/zigforum/needed-libraries?

**Comment by marnix** [comment]
Quoting iszn_11 from Discord:

> I wrote a png decoder based on stb_image.h (basically a port), including zlib/deflate decompressor. While I could just use stb, I learned a lot about pngs and deflate. Honestly, png is the easy part. It's the decompression that almost broke me :upside_down_face: .
> https://gist.github.com/iszn11/4b4ae5cb68620fde295de4053e62d566 

(Source: [Discord link](https://discord.com/channels/605571803288698900/605572611539206171/870780399293448282).)

**Comment by hdorio** [comment]
> This is a really nice article on the subject: https://www.hanshq.net/zip.html

I made a Zig version of it if anyone wants to follow along with the article with Zig code https://github.com/hdorio/hwzip.zig

I also made a Zig version of the Go package compress/flate https://github.com/ziglang/zig/pull/10552

**Comment by andrewrk** [comment]
Thanks @hdorio! Landed in 0c1df96b17fcafdb0aa973dfcb1819929719d650.
